In [1]:
import graphlab

### Load the CIFAR-10 dataset

In [2]:
image_train = graphlab.SFrame('image_train_data/')

This non-commercial license of GraphLab Create for academic use is assigned to amitha353@gmail.com and will expire on May 07, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: C:\Users\Amitha\AppData\Local\Temp\graphlab_server_1526884336.log.0


In [3]:
image_train.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 10

Data:
+-----+----------------------+------------+-------------------------------+
|  id |        image         |   label    |         deep_features         |
+-----+----------------------+------------+-------------------------------+
|  24 | Height: 32 Width: 32 |    bird    | [0.242871761322, 1.0954537... |
|  33 | Height: 32 Width: 32 |    cat     | [0.525087952614, 0.0, 0.0,... |
|  36 | Height: 32 Width: 32 |    cat     | [0.566015958786, 0.0, 0.0,... |
|  70 | Height: 32 Width: 32 |    dog     | [1.12979578972, 0.0, 0.0, ... |
|  90 | Height: 32 Width: 32 |    bird    | [1.71786928177, 0.0, 0.0, ... |
|  97 | Height: 32 Width: 32 | automobile | [1.57818555832, 0.0, 0.0, ... |
| 107 | Height: 32 Width: 32 |    dog     | [0.0, 0.0, 0.220677852631,... |
| 121 | Height: 32 Width: 32 |    bird    | [0.0, 0.23753464222, 0.0, ... |
| 136 | Height: 32 Width: 32 | automobile | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 138 | Height: 32 Width: 32 |    bird    | [0.658935725689, 0.0, 0.0,... |
+-----+----------------------+------------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [73.0, 77.0, 58.0, 71.0, 6... |
| [7.0, 5.0, 8.0, 7.0, 5.0, ... |
| [169.0, 122.0, 65.0, 131.0... |
| [154.0, 179.0, 152.0, 159.... |
| [216.0, 195.0, 180.0, 201.... |
| [33.0, 44.0, 27.0, 29.0, 4... |
| [97.0, 51.0, 31.0, 104.0, ... |
| [93.0, 96.0, 88.0, 102.0, ... |
| [35.0, 59.0, 53.0, 36.0, 5... |
| [205.0, 193.0, 195.0, 200.... |
+-------------------------------+
[10 rows x 5 columns]

### Creating a nearest neighbors model for image retrieval - using Deep Features - KNN model

In [4]:
knn_model = graphlab.nearest_neighbors.create(image_train, features = ['deep_features'],
                                             label='id')

Starting brute force nearest neighbors model training.

#### Use image retrieval model with Deep features to find similar images

In [5]:
graphlab.canvas.set_target('ipynb')
cat = image_train[18:19]
cat['image'].show()

#### Find the nearest neighbours for the cat image

In [6]:
knn_model.query(cat)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 46.848ms     |

| Done         |         | 100         | 390.605ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |       384       |      0.0      |  1   |
|      0      |       6910      | 36.9403137951 |  2   |
|      0      |      39777      | 38.4634888975 |  3   |
|      0      |      36870      | 39.7559623119 |  4   |
|      0      |      41734      | 39.7866014148 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

In [7]:
def get_images_from_ids(query_result):
    return image_train.filter_by(query_result['reference_label'], 'id')

In [8]:
cat_neighbors = get_images_from_ids(knn_model.query(cat))

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 62.471ms     |

| Done         |         | 100         | 343.726ms    |

+--------------+---------+-------------+--------------+

In [9]:
cat_neighbors.head()

Columns:
	id	int
	image	Image
	label	str
	deep_features	array
	image_array	array

Rows: 5

Data:
+-------+----------------------+-------+-------------------------------+
|   id  |        image         | label |         deep_features         |
+-------+----------------------+-------+-------------------------------+
|  384  | Height: 32 Width: 32 |  cat  | [1.04403531551, 0.0, 0.0, ... |
|  6910 | Height: 32 Width: 32 |  cat  | [1.55474901199, 0.0, 0.0, ... |
| 36870 | Height: 32 Width: 32 |  cat  | [0.240483224392, 0.0, 0.0,... |
| 39777 | Height: 32 Width: 32 |  cat  | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
| 41734 | Height: 32 Width: 32 |  cat  | [0.0, 0.0, 0.0, 0.0, 0.0, ... |
+-------+----------------------+-------+-------------------------------+
+-------------------------------+
|          image_array          |
+-------------------------------+
| [46.0, 45.0, 50.0, 47.0, 4... |
| [154.0, 133.0, 92.0, 134.0... |
| [16.0, 20.0, 19.0, 14.0, 1... |
| [145.0, 166.0, 165.0, 164.... |
| [122.0, 27.0, 34.0, 120.0,... |
+-------------------------------+
[5 rows x 5 columns]

In [10]:
cat_neighbors['image'].show()

In [11]:
car = image_train[8:9]
car['image'].show()

In [12]:
get_images_from_ids(knn_model.query(car))['image'].show()

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 93.752ms     |

| Done         |         | 100         | 390.638ms    |

+--------------+---------+-------------+--------------+

### Create a lambda to find and show the nearest neighbor images

In [13]:
show_neighbors = lambda i : get_images_from_ids(knn_model.query(image_train[i:i+1]))['image'].show()

In [14]:
show_neighbors(8)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 46.877ms     |

| Done         |         | 100         | 390.634ms    |

+--------------+---------+-------------+--------------+

In [15]:
show_neighbors(26)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 46.878ms     |

| Done         |         | 100         | 390.629ms    |

+--------------+---------+-------------+--------------+

In [16]:
show_neighbors(20)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 62.5ms       |

| Done         |         | 100         | 390.629ms    |

+--------------+---------+-------------+--------------+

In [17]:
show_neighbors(122)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 46.875ms     |

| Done         |         | 100         | 390.631ms    |

+--------------+---------+-------------+--------------+

In [18]:
show_neighbors(2000)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.0498753   | 62.499ms     |

| Done         |         | 100         | 359.377ms    |

+--------------+---------+-------------+--------------+

#### Scenario 1 : Computing summary statistics of the data:

In [19]:
image_train['label'].sketch_summary()


+------------------+-------+----------+
|       item       | value | is exact |
+------------------+-------+----------+
|      Length      |  2005 |   Yes    |
| # Missing Values |   0   |   Yes    |
| # unique values  |   4   |    No    |
+------------------+-------+----------+

Most frequent items:
+-------+------------+-----+-----+------+
| value | automobile | cat | dog | bird |
+-------+------------+-----+-----+------+
| count |    509     | 509 | 509 | 478  |
+-------+------------+-----+-----+------+


### Q1 : What’s the least common category in the training data?

### Bird (478)

### Scenario 2: Creating category-specific image retrieval models

In [20]:
image_train_automobile = image_train[image_train['label'] =='automobile']
image_train_dog = image_train[image_train['label'] =='dog']
image_train_cat = image_train[image_train['label'] == 'cat']
image_train_bird = image_train[image_train['label']=='bird']

In [21]:
knn_model_automobile = graphlab.nearest_neighbors.create(image_train_automobile,
                                                        features=['deep_features'],
                                                        label='id')
knn_model_dog = graphlab.nearest_neighbors.create(image_train_dog,
                                                 features=['deep_features'],
                                                 label='id')
knn_model_cat = graphlab.nearest_neighbors.create(image_train_cat,
                                                  features=['deep_features'],
                                                  label='id')
knn_model_bird = graphlab.nearest_neighbors.create(image_train_bird,
                                                   features=['deep_features'],
                                                   label='id')

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

In [22]:
image_test = graphlab.SFrame('image_test_data/')

In [23]:
show_neighbors_automobile = lambda i:get_images_from_ids(knn_model_automobile.query(image_test[i:i+1]))['image'].show()
show_neighbors_dog = lambda i:get_images_from_ids(knn_model_dog.query(image_test[i:i+1]))['image'].show()
show_neighbors_cats = lambda i:get_images_from_ids(knn_model_cat.query(image_test[i:i+1]))['image'].show()
show_neighbors_bird = lambda i:get_images_from_ids(knn_model_bird.query(image_test[i:i+1]))['image'].show()

### Q2 : Of the images below, which is the nearest ‘cat’ labeled image in the training data to the the first image in the test data (image_test[0:1])?

In [24]:
show_neighbors_cats(0)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 31.25ms      |

| Done         |         | 100         | 125.002ms    |

+--------------+---------+-------------+--------------+

### Q3: Of the images below, which is the nearest ‘dog’ labeled image in the training data to the the first image in the test data (image_test[0:1])?

In [25]:
show_neighbors_dog(0)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 31.242ms     |

| Done         |         | 100         | 125.002ms    |

+--------------+---------+-------------+--------------+

### Q5: For the first image in the test data, in what range is the mean distance between this image and its 5 nearest neighbors that were labeled ‘cat’ in the training data?

In [26]:
knn_model_cat.query(image_test[0:1])

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 31.251ms     |

| Done         |         | 100         | 125.003ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16289      |  34.623719208 |  1   |
|      0      |      45646      | 36.0068799284 |  2   |
|      0      |      32139      | 36.5200813436 |  3   |
|      0      |      25713      | 36.7548502521 |  4   |
|      0      |       331       | 36.8731228168 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

### Result = 36.87 (35 - 37)
### Q5: For the first image in the test data, in what range is the mean distance between this image and its 5 nearest neighbors that were labeled ‘dog’ in the training data?

In [27]:
knn_model_dog.query(image_test[0:1])

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.196464    | 15.626ms     |

| Done         |         | 100         | 125.002ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	int
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+---------------+------+
| query_label | reference_label |    distance   | rank |
+-------------+-----------------+---------------+------+
|      0      |      16976      | 37.4642628784 |  1   |
|      0      |      13387      | 37.5666832169 |  2   |
|      0      |      35867      | 37.6047267079 |  3   |
|      0      |      44603      | 37.7065585153 |  4   |
|      0      |       6094      | 38.5113254907 |  5   |
+-------------+-----------------+---------------+------+
[5 rows x 4 columns]

### Result : (38.51) 37 - 39
### Q6 : On average, is the first image in the test data closer to its 5 nearest neighbors in the ‘cat’ data or in the ‘dog’ data?

### Result : Cat (35 - 37) < Dog(37-39); Therefore Cat;
### Q7 : In what range is the accuracy of the 1-nearest neighbor classifier at classifying ‘dog’ images from the test set? 

In [28]:
image_test_automobile = image_test[image_test['label'] == 'auomobile']
image_test_dog = image_test[image_test['label'] == 'dog']
image_test_cat = image_test[image_test['label'] == 'cat']
image_test_bird = image_test[image_test['label'] == 'bird']

In [29]:
dog_automobile_neighbors = knn_model_automobile.query(image_test_dog, k =1)
dog_dog_neighbors = knn_model_dog.query(image_test_dog, k =1)
dog_cat_neighbors = knn_model_cat.query(image_test_dog, k =1)
dog_bird_neighbors = knn_model_bird.query(image_test_dog, k =1)

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 8

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 63000   | 12.3772     | 968.737ms    |

| Done         | 509000  | 100         | 1.01s        |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 8

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 63000   | 12.3772     | 1.00s        |

| Done         | 509000  | 100         | 1.10s        |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 8

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 64000   | 12.5737     | 609.377ms    |

| Done         | 509000  | 100         | 703.128ms    |

+--------------+---------+-------------+--------------+

Starting blockwise querying.

max rows per data block: 4348

number of reference data blocks: 8

number of query data blocks: 1

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 1000         | 60000   | 12.5523     | 453.125ms    |

| Done         | 478000  | 100         | 578.126ms    |

+--------------+---------+-------------+--------------+

In [30]:
dog_distance = graphlab.SFrame({'dog-automobile-distance': dog_automobile_neighbors['distance'],
                               'dog-dog-distance': dog_dog_neighbors['distance'],
                               'dog-cat-distance':dog_cat_neighbors['distance'],
                               'dog-bird-distance':dog_bird_neighbors['distance']})

In [31]:
dog_distance.head()

Columns:
	dog-automobile-distance	float
	dog-bird-distance	float
	dog-cat-distance	float
	dog-dog-distance	float

Rows: 10

Data:
+-------------------------+-------------------+------------------+------------------+
| dog-automobile-distance | dog-bird-distance | dog-cat-distance | dog-dog-distance |
+-------------------------+-------------------+------------------+------------------+
|      41.9579761457      |   41.7538647304   |  36.4196077068   |  33.4773590373   |
|      46.0021331807      |   41.3382958925   |  38.8353268874   |  32.8458495684   |
|      42.9462290692      |   38.6157590853   |  36.9763410854   |  35.0397073189   |
|      41.6866060048      |   37.0892269954   |  34.5750072914   |  33.9010327697   |
|      39.2269664935      |    38.272288694   |   34.778824791   |  37.4849250909   |
|      40.5845117698      |   39.1462089236   |  35.1171578292   |   34.945165344   |
|      45.1067352961      |    40.523040106   |  40.6095830913   |  39.0957278345   |
|      41.3221140974      |   38.1947918393   |  39.9036867306   |  37.7696131032   |
|      41.8244654995      |   40.1567131661   |  38.0674700168   |  35.1089144603   |
|      45.4976929401      |   45.5597962603   |  42.7258732951   |  43.2422832585   |
+-------------------------+-------------------+------------------+------------------+
[10 rows x 4 columns]

In [32]:
def is_dog_correct(row):
    if (row['dog-dog-distance'] < row['dog-automobile-distance'] and 
        row['dog-dog-distance'] < row['dog-cat-distance'] and row['dog-dog-distance'] < row['dog-bird-distance']):
        return 1
    else:
        return 0

In [33]:
is_dog_correct(dog_distance[0])

1

In [34]:
dog_distance.apply(is_dog_correct).sum()

678L

In [35]:
dog_distance.num_rows()

1000

In [39]:
total = float(dog_distance.num_rows())
dog_correct = float(dog_distance.apply(is_dog_correct).sum())
Result = float(dog_correct / total)
print Result * 100,'%'

67.8 %


### Result : 67.8 % (60 - 70)